internal/dataplane/parser/translate_httproute.go

```diff
@@ -279,33 +268,122 @@ func generateKongRoutesFromHTTPRouteRule(
-		if len(plugins) != 0 {
-			if r.Plugins == nil {
-				r.Plugins = make([]kong.Plugin, 0, len(plugins))
+		r.Plugins = append(r.Plugins, plugins...)
+
+		return r, nil
+	}
+
+	// TODO: implement query param matches (https://github.com/Kong/kubernetes-ingress-controller/issues/2778)
+	if len(matches[0].QueryParams) > 0 {
+		return kongstate.Route{}, fmt.Errorf("query param matches are not yet supported")
+	}
+
+	r := generateKongstateRoute(routeName, ingressObjectInfo, hostnames)
+
+	// convert header matching from HTTPRoute to Route format
+	headers, err := convertGatewayMatchHeadersToKongRouteMatchHeaders(matches[0].Headers)
+	if err != nil {
+		return kongstate.Route{}, err
+	}
+	if len(headers) > 0 {
+		r.Route.Headers = headers
+	}
+
+	// stripPath needs to be disabled by default to be conformant with the Gateway API
+	r.StripPath = kong.Bool(false)
+
+	// attach the plugins to be applied to the given route
+	if len(plugins) != 0 {
+		if r.Plugins == nil {
+			r.Plugins = make([]kong.Plugin, 0, len(plugins))
+		}
+		r.Plugins = append(r.Plugins, plugins...)
+	}
+
+	seenMethods := make(map[string]struct{})
+
+	for _, match := range matches {
+		// configure path matching information about the route if paths matching was defined
+		// Kong automatically infers whether or not a path is a regular expression and uses a prefix match by
+		// default it it is not. For those types, we use the path value as-is and let Kong determine the type.
+		// For exact matches, we transform the path into a regular expression that terminates after the value
+		if match.Path != nil {
+			path := generateKongRoutePathFromHTTPRouteMatch(match, addRegexPrefix)
+			r.Route.Paths = append(r.Route.Paths, kong.String(path))
+		}
+
+		// configure method matching information about the route if method
+		// matching was defined.
+		if match.Method != nil {
+			method := string(*match.Method)
+			if _, ok := seenMethods[method]; !ok {
+				r.Route.Methods = append(r.Route.Methods, kong.String(string(*match.Method)))
+				seenMethods[method] = struct{}{}
 			}
-			r.Plugins = append(r.Plugins, plugins...)
 		}
-		routes = append(routes, r)
 	}
 
-	return routes, nil
+	return r, nil
+}
+
+func generateKongRoutePathFromHTTPRouteMatch(match gatewayv1beta1.HTTPRouteMatch, addRegexPrefix bool) string {
+	switch *match.Path.Type {
+	case gatewayv1beta1.PathMatchExact:
+		terminated := *match.Path.Value + "$"
+		if addRegexPrefix {
+			terminated = translators.KongPathRegexPrefix + terminated
+		}
+		return terminated
+
+	case gatewayv1beta1.PathMatchPathPrefix:
+		path := *match.Path.Value
+		return path
+
+	case gatewayv1beta1.PathMatchRegularExpression:
+		path := *match.Path.Value
+		if addRegexPrefix {
+			path = translators.KongPathRegexPrefix + path
+		}
+		return path
+	}
+
+	return "" // unreachable code
+}
+
+func generateKongstateRoute(routeName string, ingressObjectInfo util.K8sObjectInfo, hostnames []*string) kongstate.Route {
```

Comment:
`hostnames` could be a variadic parameter? This way we'd avoid the need of supplying a `nil` slice in case we didn't want to provide any hostnames.

Reply:
Right now we don't have case like that. Parent function `generateKongRouteFromHTTPRouteMatches` accepts hostnames as an slice.  Functions above extract the hostname by `hostnames := getHTTPRouteHostnamesAsSliceOfStringPointers(httproute)`. 
I don't see a point.